### ISSUE TITLE
DSv4f - DSpark garbage output

### ISSUE THREAD

**Issue by mratsim** [voipmonitor/rtx6kpro#20]
As mentioned in the Discord, several people have been receiving garbed output with <| begin_of_sentence |> token and then unrelated stuff that seems to come from training data.

## Symptoms
<details>
<summary>Pics of the bug</summary>

- Call of Duty while writing a GPU compiler <img width="1377" height="2002" alt="Image" src="https://github.com/user-attachments/assets/cbff59d7-0f95-4891-9c92-6d0238fd0534" />
- TicTacToe and Snake game in React <img width="1402" height="928" alt="Image" src="https://github.com/user-attachments/assets/478c15b6-6c69-4e8c-88b9-7235ae17e515" />
- airline tickets while writing a GPU compiler <img width="1394" height="1474" alt="Image" src="https://github.com/user-attachments/assets/27d27365-4de9-4f8d-aefe-83f5719e35b3" />
- Dungeons & Dragons scenarios: <img width="1393" height="1770" alt="Image" src="https://github.com/user-attachments/assets/c5d0f04c-9f20-4e83-8325-f51893e64871" />
- Czech: <img width="1380" height="1744" alt="Image" src="https://github.com/user-attachments/assets/2d14befa-bbfb-4463-b07b-61df1fbc8479" />

</details>

## Related bugs:

- https://github.com/vllm-project/vllm-ascend/issues/9170#issuecomment-4618732340
  >   1. **`unpermute/moe_token_unpermute_tiling.h:30-34`** — sole writer of the per-token output (called on-device from `dispatch_ffn_combine_kernel.hpp:926`). Per-core token count via integer division, no min-1 guard: when a rank's local `outTokens < coreNum` (AIV cores), `tokens_core_length = outTokens/coreNum = 0` → `tokens_splited_length = min(0,600) = 0` → `tokens_splited_num = 0/0`, `tokens_splited_remain = 0%0` (integer divide-by-zero). The undefined values bound the output-emitting loop (`moe_token_unpermute.h:181-187`), so a garbage/zero number of output rows is written.
  > 
  >   2. The op **never zero-initializes** `params.ptrOutput` / combine staging (everything sized by the fixed `max_output_size=65536`), so under-filled rows on the tiny batch are read from stale HBM → NaN/Inf/degenerate logits → empty / BOS-only first response.
- https://github.com/tonyd2wild/DeepSeek-v4-Flash-DSpark-1M-NVFP4-KV-2x-DGX-Spark/commit/e83606a#diff-3d8796d2bad51a3d7286b8f2c2235d7cf7b8774001e5045f7236f8021411bd15
  - https://github.com/tonyd2wild/DeepSeek-v4-Flash-DSpark-1M-NVFP4-KV-2x-DGX-Spark/issues/3
    > ## Patch 3 — no spec placeholders on prefill chunks
    > 
    > ### Symptom
    > Resuming an existing long conversation **cold** (server restarted, or the prompt
    > fell out of the prefix cache) produced garbage at the **start** of the reply —
    > prompt echo, leaked tool/schema text, "your message was cut off"-style replies —
    > then generation recovered. Warm continuations of the same conversation were
    > clean. Reproduced deterministically with a ~98K-token prompt at `temperature 0`:
    > cold run echoed prompt content from ≈`prompt_len − 8192` (the last chunk
    > boundary) and looped; the immediately repeated (warm) run answered correctly.
    > 
    > ### Root cause
    > The DSpark async-scheduling addition in `Scheduler.update_from_output` resizes
    > the `[-1]` spec-token placeholder list to DSpark's confidence-scheduled draft
    > length. It ran for **every running, non-stopped request in the batch — including
    > requests still mid-way through a chunked prefill**. Upstream never installs spec
    > placeholders on prefill chunks: `AsyncScheduler._update_after_schedule` skips
    > `request.is_prefill_chunk`, and `Scheduler.update_draft_token_ids` explicitly
    > clears drafts for prefill chunks.
    > 
    > With a long cold prompt (multiple `max_num_batched_tokens=8192` chunks), the
    > illegal placeholders made the scheduler attach `num_speculative_tokens` spec
    > tokens to the request's **final prompt chunk**. Two corruptions follow:
    > 
    > 1. The drafts verified there were proposed from the **truncated** prompt
    >    (mid-prefill DSpark drafts — prompt-continuation predictions).
    > 2. The request was in the previous step's **discard set** (mid-prefill), so the
    >    worker excludes it from `prev_req_id_to_index`; `_prepare_input_ids` then
    >    never scatters real draft ids over the `-1` placeholders, and the target
    >    forward for the final chunk sees invalid token ids at the spec positions.
    > 
    > Either way the first sampled tokens of the reply are conditioned on a corrupted
    > prompt tail — visible as prompt echo / garble at the start of a cold resume,
    > recovering once pure decode steps take over. (Earlier this corrupted whole
    > conversations; Patches 1/2/2b fixed the persistent-KV side, leaving only the
    > cold-start window.)
    > 
    > ### Fix
    > `vllm/v1/core/sched/scheduler.py` (`update_from_output`): only resize spec
    > placeholders for requests the AsyncScheduler itself would give placeholders —
    > `new_token_ids` non-empty, `not request.is_prefill_chunk`, and
    > `request.status == RequestStatus.RUNNING` (a preempted request must keep its
    > cleared spec list).
    > 
    > ### Validation
    > Same ~98K-token cold-resume prompt at `temperature 0` after the fix: the cold
    > first reply is correct (no echo, no cut-off complaint), and matches the warm
    > rerun. Short direct prompts and the deterministic `NVFP4 DSPARK OK` check are
    > unchanged.

## Suspicions

DSpark is the first speculative decoding with dynamically predicted tokens since the confidence head can choose to predict only 3 tokens out of 5. It seems very plausible that with DSpark=5, but 3 predicted, the extra 2 placeholders are uninitialized at <| begin_of_sentence |> token. Then this corrupts the model and it starts to regurgitate training data.

## Tentative patch

I tried making a patch against `voipmonitor/vllm:eldritch-enlightenment-v45c1582-b12xf3686b5-pc1441b5-cu132-20260704` image (v9), so that I can iterate without having to rebuild Dockerfile.

It can be applied via
```bash
# Get current script dir
DIR=$(realpath "$(dirname "${BASH_SOURCE[0]}")")

# Docker / podman command
podman run \
    `# [...]` \
    `# Mount volume to Docker/podman run` \
    -v "${DIR}/run_vllm_DeepSeek-V4-Flash.dspark-v9.patch.py":/app/dspark_scheduler_patch.py:ro \
    `# [...]` \
    --entrypoint /bin/bash \
    "${IMAGE}" \
        -lc 'unset NCCL_GRAPH_FILE NCCL_GRAPH_DUMP_FILE VLLM_B12X_MLA_EXTEND_MAX_CHUNKS; python3 /app/dspark_scheduler_patch.py && exec vllm serve "$@"' -- \
            "${MODEL}" \
            `# Networking` \
            --host 0.0.0.0 \
            --port "${VLLM_PORT}" \
```

**It doesn't work yet** but it should help having a faster testing loop without rebuilding vllm or compilingflash-attention.
<details>

<summary>Patch</summary>

```py
"""
Monkey-patch for vLLM Scheduler.update_from_output.

Applies Patch 3 from tonyd2wild's DSpark fork (commit e83606a):
Guard spec-placeholder resize to only apply on genuine decode steps.

Root cause: the DSpark async-scheduling code in Scheduler.update_from_output
resized the [-1] spec-token placeholder list for EVERY running request,
including ones mid chunked-prefill. Upstream never installs spec placeholders
on prefill chunks, so on a long cold prompt (multiple 8192-tok chunks) this
attached spec tokens to the final prompt chunk, corrupting the prompt tail.

Fix: only resize placeholders on genuine decode steps -- new_token_ids non-empty,
not request.is_prefill_chunk, and status == RUNNING (preempted requests keep a
cleared spec list).

This version does NOT require adding `draft_token_lengths` to ModelRunnerOutput.
Instead, it detects DSpark via vllm_config.speculative_config.method and computes
the draft length from `len(eo.new_token_ids) - 1` (the number of accepted drafts
without the verified token), which is already available in update_from_output.

The resize works because the next call to Scheduler.schedule() reads the
spec_token_ids to determine how many draft tokens to schedule. If we resize
to [-1] * num_accepted, the next schedule will schedule exactly num_accepted
tokens instead of num_spec_tokens.
"""

import logging

logger = logging.getLogger(__name__)

_PATCH_MARKER = "_dspark_scheduler_patch_applied_v3"


def apply_patch():
    """Apply the scheduler patch to fix DSpark cold-resume garble."""
    try:
        from vllm.v1.core.sched.scheduler import Scheduler, RequestStatus
    except ImportError as e:
        logger.error("Failed to import vLLM scheduler: %s", e)
        return

    # Avoid double-patching
    if getattr(Scheduler, _PATCH_MARKER, False):
        logger.info("DSpark scheduler patch already applied, skipping.")
        return

    original_update_from_output = Scheduler.update_from_output

    def patched_update_from_output(self, scheduler_output, model_runner_output):
        # ---------------------------------------------------------------
        # 1. Determine if this is a DSpark run
        # ---------------------------------------------------------------
        is_dspark = False
        spec_config = getattr(self.vllm_config, "speculative_config", None)
        if spec_config is not None:
            method = getattr(spec_config, "method", None)
            if method == "dspark":
                is_dspark = True

        # ---------------------------------------------------------------
        # 2. Run the original logic first
        # ---------------------------------------------------------------
        outputs = original_update_from_output(
            self, scheduler_output, model_runner_output
        )

        if not is_dspark:
            return outputs

        # ---------------------------------------------------------------
        # 3. Apply the guarded spec-placeholder resize for decode steps
        # ---------------------------------------------------------------
        # The original code (without this patch) does NOT resize spec_token_ids
        # at all. We add it here by iterating over processed requests and
        # setting spec_token_ids = [-1] * draft_len for requests that satisfy:
        #   - had new_token_ids this step (decode, not prefill)
        #   - not request.is_prefill_chunk
        #   - request.status == RequestStatus.RUNNING
        #
        # draft_len = len(new_token_ids) - 1 = num_accepted (drafts accepted
        # by the target model, not counting the verified token).
        num_scheduled_tokens = scheduler_output.num_scheduled_tokens
        for req_id, num_tokens_scheduled in num_scheduled_tokens.items():
            request = self.requests.get(req_id)
            if request is None or request.is_finished():
                continue

            # Determine if this request produced new tokens in this step
            # by scanning the EngineCoreOutputs.
            has_new_tokens = False
            for engine_outputs in outputs.values():
                for eo in engine_outputs:
                    if eo.request_id == req_id and eo.new_token_ids:
                        has_new_tokens = True
                        break
                if has_new_tokens:
                    break

            if (
                has_new_tokens
                and not request.is_prefill_chunk
                and request.status == RequestStatus.RUNNING
            ):
                # Compute draft length from the number of accepted tokens.
                # For each decode step with DSpark:
                #   new_token_ids = [verified_token] + [accepted_drafts]
                # So len(new_token_ids) - 1 = num_accepted drafts.
                for engine_outputs in outputs.values():
                    for eo in engine_outputs:
                        if eo.request_id == req_id:
                            token_count = len(eo.new_token_ids)
                            if token_count > 0:
                                draft_len = max(
                                    0, min(token_count - 1, self.num_spec_tokens)
                                )
                                request.spec_token_ids = [-1] * draft_len
                            break
                    if has_new_tokens:
                        break

        return outputs

    Scheduler.update_from_output = patched_update_from_output
    setattr(Scheduler, _PATCH_MARKER, True)
    logger.info(
        "Applied DSpark scheduler patch: guarded spec-placeholder resize (cold-start fix)"
    )


if __name__ == "__main__":
    apply_patch()
```

</details>

**Comment by officiallymarky** [comment]
I have been seeing this frequently, it works for a while, then it was just dump the template nonsense and random training data.

**Comment by mratsim** [comment]
Closing, not seeing this in v9 or v10 anymore

**Comment by voipmonitor** [comment]
r27 is the current 0731 DS4 release and retains fixed K5 as the default. It passed short batch-correctness and concurrent long-prefill gates, but rare production BOS collapse is still tracked separately in #53 and is not declared fixed.

Image and runbook:

- `voipmonitor/vllm:gilded-gnosis-v20-vllm966d57c-sibbbdccc-fi801d57a-cu132-20260803-r27`
- https://github.com/local-inference-lab/rtx6kpro/blob/master/models/ds4dspark-v20.md

Use #53 for any new batch-wide BOS reproduction so launch configuration, context/concurrency, and acceptance telemetry remain in one current tracker.
